pennylane/qinfo/classical_fisher.py

```diff
@@ -0,0 +1,143 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Differentiable classical fisher information"""
+
+import pennylane as qml
+import pennylane.numpy as pnp
+
+import jax
+import torch
+import tensorflow as tf
+import tensorflow.python.ops.numpy_ops.np_config as np_config
+
+
+def torch_jac(circ):
+    def wrapper(params):
+        return torch.autograd.functional.jacobian(circ, (params))
+
+    return wrapper
+
+
+def tf_jac(circ):
+    def wrapper(params):
+        with tf.GradientTape() as tape:
+            loss = circ(params)
+        return tape.jacobian(loss, params)
+
+    return wrapper
+
+
+def CFIM(qnode):
+    """Computing the classical fisher information matrix (CFIM) using the jacobian of the output probabilities
+    as described in eq. (15) in https://arxiv.org/abs/2103.15191
+    """
+    new_qnode = qml.transforms._make_probs(
+        qnode, post_processing_fn=lambda x: qml.math.squeeze(qml.math.stack(x))
+    )
+
+    interface = qnode.interface
+
+    if interface == "jax":
+        jac = jax.jacobian(new_qnode)
+    if interface == "torch":
+        jac = torch_jac(new_qnode)
+    if interface == "autograd":
+        jac = qml.jacobian(new_qnode)
+    if interface == "tf":
+        jac = tf_jac(new_qnode)
+        np_config.enable_numpy_behavior()  # this allows for the manipulations in _compute_cfim with tensors
+
+    def wrapper(*args, **kwargs):
+
+        j = jac(*args, **kwargs)
+        p = qnode(*args, **kwargs)
+
+        cfim = _compute_cfim(p, j, interface)
+        return cfim
+
+    return wrapper
+
+
+def _compute_cfim(p, dp, interface):
+    """Computes the (num_params, num_params) classical fisher information matrix from the probabilities and its derivatives
+    I.e. it computes :math:`CFIM_{ij} = \sum_\ell (\partial_i p_\ell) (\partial_i p_\ell) / p_\ell`
+    """
+
+    # Compute 1/p for p!=0, else 0
```

Comment:
This part somehow escalated into this mess below by writing the tests for the different interfaces. I'm gonna try to rewrite this part.

Reply:
Same comment as before, you create a `qml.math` function by registering the function in `single_dispatch` for all interfaces.